### PR TITLE
L2VPN: Implement serilization of EVPN-VPWS

### PR DESCRIPTION
**Pull request by vidister**
see #3

### CODE REVIEW

File: cosmo/serializer.py [hunk 1]
```diff
@@ -87,11 +87,21 @@ def _get_unit(self, iface):
...
+        if l2vpn:
+            if l2vpn['type'] == "VPWS":
+                if iface.get("untagged_vlan"):
+                    unit_stub["encapsulation"] = "vlan-ccc"
+                else:
+                    unit_stub["encapsulation"] = "ethernet-ccc"
+            elif l2vpn['type'] in ["MPLS_EVPN", "VXLAN_EVPN"]:
                 unit_stub["encapsulation"] = "vlan-bridge"
+
+            if unit_stub["encapsulation"]:
```
Comment (johannwagner) [inline cosmo/serializer.py:104]: What do you want to check here?
Probably something like `if l2vpn['type'] in ['x','y']` or another variable with a describing name would be helpful.

Comment (johannwagner) [reply cosmo/serializer.py:104]: Also I think, this variable is set every time for every valid L2VPN we support...

File: cosmo/serializer.py [hunk 2]
```diff
@@ -233,6 +243,28 @@ def serialize(self):
...
+                    "instance_type": "evpn-vpws",
+                    "protocols": {
+                        "evpn": {
+                            "interfaces": {
+                                i["name"]: {
+                                    "vpws_service_id": {
+                                        "local": int(i['id']),
+                                        "remote": int([termination["assigned_object"]["id"] for termination in l2vpn["terminations"] if termination["assigned_object"]["id"] != i['id']][0]),
+                                    },
+                                } for i in l2vpn["interfaces"]
```
Comment (johannwagner) [inline cosmo/serializer.py:261]: You might want to to generatet this before the object declaration and use some auxiliary varialbes for better and more readable naming. Especially the `remote` key in `vpws_service_id`.. :laughing:

File: cosmo/__main__.py [hunk 3]
```diff
@@ -60,6 +60,9 @@ def noop(*args, **kwargs):
     for l2vpn in cosmo_data["l2vpn_list"]:
         if not l2vpn["name"].startswith("WAN: "):
             continue
+        if l2vpn['type'] == "VPWS" and len(l2vpn['terminations']) != 2:
```
Comment (johannwagner) [inline cosmo/__main__.py:63]: You also want to check, that VPWS L2VPNs only contain interfaces?

**Review by johannwagner — CHANGES_REQUESTED** [review]

File: cosmo/__main__.py (hunk 3 above)
Comment (vidister) [reply cosmo/__main__.py:63]: Using VLANs to expand to interfaces is totally valid, even for VPWS setups.
But I get that we can't guarantee that we only have two endpoints then. What do you think? Shall we expect explicit interface configuration for VPWS tunnels?

Comment (johannwagner) [reply cosmo/__main__.py:63]: Netbox only allows two terminations for VPWS, but I am not sure, if those two terminations also can be two different VLANs. 

Therefore, I would postulate, that if we want to connect two interfaces, we should use two interface terminations and and if we want to connect two VLANs, we should use VLAN interfaces - but I am not sure, if this is a thing in VPWS.

Comment (vidister) [reply cosmo/__main__.py:63]: > Netbox only allows two terminations for VPWS

this is gonna be fun when we implement Multi-Homed Ethernet Segments for VPWS. But that's our problem for later.

Anyways, VPWS circuits on VLANs are a thing. But not to confuse two things:
What we are doing here is not connecting a VPWS circuit to a VLAN bridge domain, but expanding VLANs to Interfaces, to connect a circuit to it. (Even though in an EVPN (non-VPWS) circuit we implicitly configure a bridge domain by doing so, so here that might make sense)

TL;DR: All this can be a bit confusing. So I'd follow your suggestion and only allow interface terminations on VPWS circuits - as it's also possible to terminate a VPWS circuit on a VLAN sub-interface.

**Review by johannwagner — APPROVED** [review]
